x/authz/client/cli/tx.go

```diff
@@ -0,0 +1,197 @@
+package cli
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/spf13/cobra"
+
+	"github.com/cosmos/cosmos-sdk/client"
+	"github.com/cosmos/cosmos-sdk/client/flags"
+	"github.com/cosmos/cosmos-sdk/client/tx"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/cosmos-sdk/version"
+	authclient "github.com/cosmos/cosmos-sdk/x/auth/client"
+	"github.com/cosmos/cosmos-sdk/x/authz/types"
+)
+
+// GetTxCmd returns the transaction commands for this module
+func GetTxCmd() *cobra.Command {
+	AuthorizationTxCmd := &cobra.Command{
+		Use:                        types.ModuleName,
+		Short:                      "Authorization transactions subcommands",
+		Long:                       "Authorize and revoke access to execute transactions on behalf of your address",
+		DisableFlagParsing:         true,
+		SuggestionsMinimumDistance: 2,
+		RunE:                       client.ValidateCmd,
+	}
+
+	AuthorizationTxCmd.AddCommand(
+		NewCmdGrantAuthorization(),
+		NewCmdRevokeAuthorization(),
+		NewCmdExecAuthorization(),
+	)
+
+	return AuthorizationTxCmd
+}
+
+func NewCmdGrantAuthorization() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "grant [grantee_address] [msg-type] [limit] --from=[granter]",
+		Short: "Grant authorization to an address",
+		Long: strings.TrimSpace(
+			fmt.Sprintf(`Grant authorization to an address to execute a transaction on your behalf:
+
+Examples:
+ $ %s tx %s grant cosmos1skjw.. %s 1000stake --from=cosmos1skl..
+ $ %s tx %s grant cosmos1skjw.. /cosmos.gov.v1beta1.Msg/Vote --from=cosmos1sk..
+	`, version.AppName, types.ModuleName, types.SendAuthorization{}.MethodName(), version.AppName, types.ModuleName),
+		),
+		Args: cobra.RangeArgs(2, 3),
+		RunE: func(cmd *cobra.Command, args []string) error {
+			clientCtx, err := client.GetClientTxContext(cmd)
+			if err != nil {
+				return err
+			}
+			grantee, err := sdk.AccAddressFromBech32(args[0])
+			if err != nil {
+				return err
+			}
+
+			msgType := args[1]
+
+			var authorization types.Authorization
+			if msgType == (types.SendAuthorization{}.MethodName()) {
+				limit, err := sdk.ParseCoinsNormalized(args[2])
+				if err != nil {
+					return err
+				}
+				authorization = &types.SendAuthorization{
+					SpendLimit: limit,
+				}
+			} else {
+				authorization = types.NewGenericAuthorization(msgType)
```

Comment:
we need to validate if given message type exists, it's unclear to me what happens if method name is wrong.

Reply:
Yes, it would be nice to have this validation but I cannot think a clean way to handle this. Any suggestions @aaronc @amaurymartiny @robert-zaremba @sahith-narahari ? Also, this will sit in client, so someone can use a new client and still send rubbish data. If there's a way to add this check in msg server, that would work best

Reply:
Had a discussion with @amaurymartiny offline and found a way to validate this on server side.